ssl/record/methods/tls_common.c

```diff
@@ -284,27 +290,9 @@ int tls_default_read_n(OSSL_RECORD_LAYER *rl, size_t n, size_t max, int extend,
 
     if (!extend) {
         /* start with empty packet ... */
-        if (left == 0) {
+        if (left == 0)
             rb->offset = align;
-        } else if (align != 0 && left >= SSL3_RT_HEADER_LENGTH) {
-            /*
-             * check if next packet length is large enough to justify payload
-             * alignment...
-             */
-            pkt = rb->buf + rb->offset;
-            if (pkt[0] == SSL3_RT_APPLICATION_DATA
-                    && (pkt[3] << 8 | pkt[4]) >= 128) {
-                /*
-                 * Note that even if packet is corrupted and its length field
-                 * is insane, we can only be led to wrong decision about
-                 * whether memmove will occur or not. Header values has no
-                 * effect on memmove arguments and therefore no buffer
-                 * overrun can be triggered.
-                 */
-                memmove(rb->buf + align, pkt, left);
-                rb->offset = align;
-            }
-        }
```

Comment:
Does this special case no longer apply?

Reply:
This memmove was originally intended as an optimisation. It ensures that the payload data bytes are properly aligned for maximum assembler performance. There are two cases to consider: one where clearold is 0 and one where clearold is 1.

In the case that clearold is 0 this means "don't invalidate data earlier in the buffer because we're still using it". So this memove is just wrong in that instance. We must not move the data to an earlier position in the buffer because we're still using it. This actually was causing a problem in the pipelining case because we have multiple pipelines of data all using the same buffer - so this memmove was resulting in "bad record mac" error.

In the case that clearold is 1 it would still be a valid optimisation - but is entirely redundant  because of the memmove that occurs a few lines later. This *always* occurs if clearold is 1 and the data is not already aligned.

So, since this code is either wrong or redundant (depending on the value of clearold) it can safely be removed completely.